### PR TITLE
🔍  Add clang-tidy.

### PR DESCRIPTION
**Pull request by thorstenhater**
To improve our Static Analysis story:
- add .clang-tidy with some salient checks.
- add an ignore-list for the checks I think are less useful in Arbor's code.
- ignore -- for now -- some dark places: ext, text, and modcc
- fix many of the issues found

I have used PVS Studio in the past (and I still think it's the better tool) but they require
some weird contortions wrt README acknowledgements.

I found some potential bugs
- use-after-move in nml_parse_morph 
- (dubious) use-after-move in locked_ostream that seems unfixable
- incorrect erase/remove_if in communicator
- no check against self-assignment in any_ptr
- some incorrect/redundant `moves`
- missing includes

### CODE REVIEW

**Comment by boeschf** [comment]
Clang-tidy cannot explore all code paths (because it uses parts of the compiler itself) - in particular, it will not look at code that is unreachable due to a macro not being defined (ifdefs etc). This affects, for example,
- customizations for other compilers
- GPU backend (and obviously cuda files)

This is not a critique, but we should keep this in mind. Btw, did you compile the tidied code with GPU support?

File: arbor/include/arbor/util/compat.hpp [hunk 1]
```diff
@@ -9,7 +9,7 @@
 
 namespace compat {
 
-constexpr bool using_intel_compiler(int major=0, int minor=0, int patchlevel=0) {
+constexpr bool using_intel_compiler(int=0, int=0, int=0) {
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/include/arbor/util/compat.hpp:12]: Who reads _headers_ anyhow? :D

File: arborio/xml.hpp [hunk 2]
```diff
@@ -43,33 +43,37 @@ T get_attr(const xml_node& n,
...
-             if (j==npos) break;
-             r += ",";
-             i = j+1;
-         }
-         r += ")";
-         return r;
-     }
+
+    auto constexpr squote = '\'';
+    auto constexpr dquote = '\'';
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arborio/xml.hpp:48]: Oh yes, my mistake!

File: arborio/neurolucida.cpp [hunk 3]
```diff
@@ -285,7 +285,7 @@ parse_hopefully<zsmear> parse_zsmear(asc::lexer& L) {
     return s;
 }
 
-#define PARSE_ZSMEAR(L, X) if (auto rval__ = parse_zsmear(L)) X=*rval__; else return FORWARD_PARSE_ERROR(rval__.error());
+#define PARSE_ZSMEAR(L, X)
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arborio/neurolucida.cpp:288]: No, it's used once, so I splat it into its use site and promptly forgot to delete!

File: arbor/include/arbor/recipe.hpp [hunk 4]
```diff
@@ -22,7 +22,8 @@ struct probe_info {
     probe_info(probe_info&&) = default;
 
     // Implicit ctor uses tag of zero.
-    template <typename X>
+    template <typename X,
+              typename = std::enable_if<!std::is_same_v<X, probe_tag>>> // otherwise, we'd hide ctors above
     probe_info(X&& x, probe_tag tag = 0):
         tag(tag), address(std::forward<X>(x)) {}
 };
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/include/arbor/recipe.hpp:29]: Sure, that's a better phrasing!

File: arbor/include/arbor/morph/isometry.hpp [hunk 5]
```diff
@@ -33,11 +33,11 @@ struct ARB_ARBOR_API isometry {
     double tx_ = 0, ty_ = 0, tz_ = 0;
 
     isometry(quaternion q, double tx, double ty, double tz):
-        q_(std::move(q)), tx_(tx), ty_(ty), tz_(tz) {}
+        q_(q), tx_(tx), ty_(ty), tz_(tz) {}
```
[the remark this answers is not in the document]
Comment (thorstenhater) [reply arbor/include/arbor/morph/isometry.hpp:36]: Well, `quaternion` is `{double, double, double, double}` thus `move` has no effect.

[the remark this answers is not in the document]
Comment (boeschf) [reply arbor/include/arbor/morph/isometry.hpp:36]: ok

**Comment by thorstenhater** [comment]
Diverged too much, closing.

**Pull request closed without merging**